dali/pipeline/operators/arithmetic/arithmetic_meta.h

```diff
@@ -0,0 +1,345 @@
+    case ArithmeticOp::mul:
+    case ArithmeticOp::div:
+    case ArithmeticOp::mod:
+      return 2;
+    default:
+      return -1;
+  }
+}
+
+/**
+ * @brief Type promotion rules
+ *
+ * @tparam L
+ * @tparam R
+ *
+ * The rules are:
+ * T      + T      = T
+ * floatX + T      = floatX           (where T is not a float)
+ * floatX + floatY = float(max(X, Y))
+ * intX   + intY   = int(max(X, Y))
+ * uintX  + uintY  = uint(max(X, Y))
+ * intX   + uintY  = int2Y            (if X <= Y)
+ * intX   + uintY  = intX             (if X > Y)
+ *
+ * Operating on integer types never leaves the integer types, and caps at [u]int64
+ *
+ * @return type - a result type of binary arithmetic operation between L and R
+ */
+template <typename L, typename R>
+struct binary_op_promotion;
+
+template <typename T>
+struct binary_op_promotion<T, T> {
+  using type = T;
+};
+
+#define REGISTER_TYPE_PROMOTION(Arg1, Arg2, Result)              \
+template <>                                                      \
+struct binary_op_promotion<Arg1, Arg2> { using type = Result; }; \
+template <>                                                      \
+struct binary_op_promotion<Arg2, Arg1> { using type = Result; }
+
+
+REGISTER_TYPE_PROMOTION(float,   float16, float);
+REGISTER_TYPE_PROMOTION(double,  float16, double);
+
+REGISTER_TYPE_PROMOTION(double,  float, double);
+
+REGISTER_TYPE_PROMOTION(int8_t,   float16, float16);
+REGISTER_TYPE_PROMOTION(uint8_t,  float16, float16);
+REGISTER_TYPE_PROMOTION(int16_t,  float16, float16);
+REGISTER_TYPE_PROMOTION(uint16_t, float16, float16);
+REGISTER_TYPE_PROMOTION(int32_t,  float16, float16);
+REGISTER_TYPE_PROMOTION(uint32_t, float16, float16);
+REGISTER_TYPE_PROMOTION(int64_t,  float16, float16);
+REGISTER_TYPE_PROMOTION(uint64_t, float16, float16);
+
+REGISTER_TYPE_PROMOTION(int8_t,   float, float);
+REGISTER_TYPE_PROMOTION(uint8_t,  float, float);
+REGISTER_TYPE_PROMOTION(int16_t,  float, float);
+REGISTER_TYPE_PROMOTION(uint16_t, float, float);
+REGISTER_TYPE_PROMOTION(int32_t,  float, float);
+REGISTER_TYPE_PROMOTION(uint32_t, float, float);
+REGISTER_TYPE_PROMOTION(int64_t,  float, float);
+REGISTER_TYPE_PROMOTION(uint64_t, float, float);
+
+REGISTER_TYPE_PROMOTION(int8_t,   double, double);
+REGISTER_TYPE_PROMOTION(uint8_t,  double, double);
+REGISTER_TYPE_PROMOTION(int16_t,  double, double);
+REGISTER_TYPE_PROMOTION(uint16_t, double, double);
+REGISTER_TYPE_PROMOTION(int32_t,  double, double);
+REGISTER_TYPE_PROMOTION(uint32_t, double, double);
+REGISTER_TYPE_PROMOTION(int64_t,  double, double);
+REGISTER_TYPE_PROMOTION(uint64_t, double, double);
+
+REGISTER_TYPE_PROMOTION(uint8_t,  int8_t, int16_t);
+REGISTER_TYPE_PROMOTION(int16_t,  int8_t, int16_t);
+REGISTER_TYPE_PROMOTION(uint16_t, int8_t, int32_t);
+REGISTER_TYPE_PROMOTION(int32_t,  int8_t, int32_t);
+REGISTER_TYPE_PROMOTION(uint32_t, int8_t, int64_t);
+REGISTER_TYPE_PROMOTION(int64_t,  int8_t, int64_t);
+REGISTER_TYPE_PROMOTION(uint64_t, int8_t, int64_t);
```

Comment:
int128_t? :-)

Reply:
I had that commented here, but @mzient complained :P